test/unittest/rtps/writer/ReaderProxyTests.cpp

```diff
@@ -331,6 +331,49 @@ TEST(ReaderProxyTests, process_nack_frag_multiple_fragments_different_windows_te
             TOTAL_NUMBER_OF_FRAGMENTS + 1u), TOTAL_NUMBER_OF_FRAGMENTS + 1u);
 }
 
+TEST(ReaderProxyTests, has_been_delivered_test)
+{
+    StatefulWriter writer_mock;
+    WriterTimes w_times;
+    RemoteLocatorsAllocationAttributes alloc;
+    ReaderProxy rproxy(w_times, alloc, &writer_mock);
+
+    CacheChange_t seq1;
+    CacheChange_t seq2;
+    seq1.sequenceNumber = {0, 1};
+    seq2.sequenceNumber = {0, 2};
+
+    ReaderProxyData reader_attributes(0, 0);
+    reader_attributes.m_qos.m_reliability.kind = RELIABLE_RELIABILITY_QOS;
+    rproxy.start(reader_attributes);
+
+    // Change 1 has been acknowledged and has been removed from the ReaderProxy
+    // Change 2 has been delivered
+    // Change 3 is pending to be delivered
+    rproxy.add_change(ChangeForReader_t(&seq1), true, false);
+    rproxy.add_change(ChangeForReader_t(&seq2), true, false);
+    rproxy.from_unsent_to_status(seq2.sequenceNumber, ACKNOWLEDGED, false, true);
+
+    // Change delivered
+    bool found = false;
+    EXPECT_FALSE(rproxy.has_been_delivered(seq1.sequenceNumber, found));
+    EXPECT_TRUE(found);
+    // Change not delivered
+    found = false;
+    EXPECT_TRUE(rproxy.has_been_delivered(seq2.sequenceNumber, found));
+    EXPECT_TRUE(found);
+    // Change acknowledged
+    rproxy.from_unsent_to_status(seq1.sequenceNumber, ACKNOWLEDGED, false, true);
+    rproxy.acked_changes_set(seq1.sequenceNumber);
+    found = false;
+    EXPECT_TRUE(rproxy.has_been_delivered(seq1.sequenceNumber, found));
+    EXPECT_FALSE(found);
+    // Change not added
+    found = false;
+    EXPECT_FALSE(rproxy.has_been_delivered({0, 4}, found));
+    EXPECT_FALSE(found);
```

Comment:
Let's try to follow the logic that makes more sense,

```suggestion
    auto expect_result = [&rproxy](SequenceNumber_t seq, bool delivered, bool should_be_found)
    {
        bool found;
        EXPECT_EQ(delivered, rproxy.has_been_delivered(seq, found));
        EXPECT_EQ(should_be_found, found);
    };
    
    // Add changes 1 and 2
    rproxy.add_change(ChangeForReader_t(&seq1), true, false);
    rproxy.add_change(ChangeForReader_t(&seq2), true, false);

    // None of them has been delivered
    expect_result(seq1.sequenceNumber, false, true);
    expect_result(seq2.sequenceNumber, false, true);

    // Change 1 is sent
    rproxy.from_unsent_to_status(seq1.sequenceNumber, UNACKNOWLEDGED, false, true);

    // Only change 1 has been delivered. Both are found.
    expect_result(seq1.sequenceNumber, true, true);
    expect_result(seq2.sequenceNumber, false, true);

    // Change 1 is acknowledged
    rproxy.acked_changes_set(seq1.sequenceNumber + 1);

    // Only change 1 has been delivered. Only change 2 is found.
    expect_result(seq1.sequenceNumber, true, false);
    expect_result(seq2.sequenceNumber, false, true);

    // Change in the future should return not delivered and not found
    expect_result({0, 3}, false, false);
```

Reply:
There is a problem with the suggested test flow. If the lowest change is acknowledged, then it is removed from the ReaderProxy. That's the reason why the change 2 is the one that is marked as delivered and not the first one.